core/data/src/main/kotlin/com/google/samples/apps/nowinandroid/core/data/util/TimeZoneMonitor.kt

```diff
@@ -0,0 +1,94 @@
+import android.content.Intent
+import android.content.IntentFilter
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import androidx.core.os.trace
+import com.google.samples.apps.nowinandroid.core.network.Dispatcher
+import com.google.samples.apps.nowinandroid.core.network.NiaDispatchers.IO
+import com.google.samples.apps.nowinandroid.core.network.di.ApplicationScope
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.shareIn
+import java.time.ZoneId
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Utility for reporting current timezone the device has set.
+ * It always emits at least once with default setting and then for each TZ change.
+ */
+interface TimeZoneMonitor {
+    val currentZoneId: Flow<ZoneId>
+}
+
+@Singleton
+internal class TimeZoneBroadcastMonitor @Inject constructor(
+    @ApplicationContext private val context: Context,
+    @ApplicationScope appScope: CoroutineScope,
+    @Dispatcher(IO) private val ioDispatcher: CoroutineDispatcher,
+) : TimeZoneMonitor {
+
+    override val currentZoneId: SharedFlow<ZoneId> =
+        callbackFlow<ZoneId> {
+            // Send the default time zone first.
+            trySend(ZoneId.systemDefault())
+
+            // Registers BroadcastReceiver for the TimeZone changes
+            val receiver = object : BroadcastReceiver() {
+                override fun onReceive(context: Context, intent: Intent) {
+                    if (intent.action != Intent.ACTION_TIMEZONE_CHANGED) return
+
+                    val zoneIdFromIntent = if (VERSION.SDK_INT < VERSION_CODES.R) {
+                        null
+                    } else {
+                        // Starting Android R we also get the new TimeZone.
+                        intent.getStringExtra(Intent.EXTRA_TIMEZONE)?.let { zoneId ->
+                            // We need to convert it from java.util.Timezone to java.time.ZoneId
+                            ZoneId.of(zoneId, ZoneId.SHORT_IDS)
+                        }
+                    }
+
+                    // If there isn't a zoneId in the intent, fallback to the systemDefault, which should also reflect the change
+                    trySend(zoneIdFromIntent ?: ZoneId.systemDefault())
+                }
+            }
+
+            trace("TimeZoneBroadcastReceiver.register") {
+                context.registerReceiver(receiver, IntentFilter(Intent.ACTION_TIMEZONE_CHANGED))
+            }
+
+            awaitClose {
+                context.unregisterReceiver(receiver)
+            }
+        }
+            .flowOn(ioDispatcher)
```

Comment:
A `callbackFlow` has a fixed buffer size by default, but since we only care about the most recent value I think we can call `.conflate()` here to ensure that the `trySend`s never fail.

```suggestion
            .conflate()
            .flowOn(ioDispatcher)
```